fc-community-backend/apps/api/src/factchain-core/nftService.ts

```diff
@@ -0,0 +1,134 @@
+import axios from "axios";
+import FormData from "form-data";
+import Replicate from "replicate";
+import { Note } from "./types";
+
+export class NFTService {
+
+  static generateNoteImage = async (note: Note, count = 0): Promise<string> => {
+    if (note.content === undefined) {
+      throw new Error("Can't generate image from empty note!");
+    }
+    const replicate = new Replicate({
+      auth: process.env.REPLICATE_API_TOKEN,
+    });
+
+    try {
+      const replicateUrlObj = await replicate.run(
+        "stability-ai/sdxl:c221b2b8ef527988fb59bf24a8b97c4561f1c671f73bd389f866bfb27c061316",
+        {
+          input: {
+            width: 768,
+            height: 768,
+            prompt: note.content,
+            refine: "expert_ensemble_refiner",
+            scheduler: "K_EULER",
+            lora_scale: 0.6,
+            num_outputs: 1,
+            guidance_scale: 7.5,
+            apply_watermark: false,
+            high_noise_frac: 0.8,
+            negative_prompt: "",
+            prompt_strength: 0.8,
+            num_inference_steps: 25,
+          },
+        },
+      );
+
+      const replicateUrl = (replicateUrlObj as string[])[0];
+      console.log(`replicate URL: ${replicateUrl}`);
+      return replicateUrl;
+    } catch (error) {
+      // retry twice if the note generated NSFW image
+      // experimentally, chance of success are close to zero after two failures for NSFW content
+      if (
+        error instanceof Error &&
+        error.message.includes("NSFW") &&
+        count < 3
+      ) {
+        return NFTService.generateNoteImage(note, count + 1);
+      } else {
+        throw error;
+      }
+    }
+  };
+
+  static uploadImageToPinata = async (
+    note: Note,
+    replicateUrl: string,
+  ): Promise<string> => {
+    const imageStream = (
+      await axios.get(replicateUrl, { responseType: "stream" })
+    ).data;
+    const data = new FormData();
+    data.append("file", imageStream, {
+      filepath: `${note.content!.substring(0, 16).replaceAll(" ", "_")}.png`, // required by the Pinata API
```

Comment:
So two notes with the same start will have the same filename? That seems brittle, maybe use a random uuid instead?